### PR TITLE
Quick fix in CryspyCalculator for objects of identical names

### PR DESCRIPTION
**Pull request by wardsimon**
Previously an error would be thrown when multiple experiments or phases have the same name. Now, if they have the same name a numeric addition will be added to the new objects name.

### CODE REVIEW

File: easyInterface/Diffraction/Calculators/CryspyCalculator.py [hunk 1]
```diff
@@ -135,14 +136,31 @@ def setExpsDefinition(self, exp_path: str):
...
         if self._cryspy_obj.crystals is not None:
             self._cryspy_obj.experiments[0].phase.items[0] = (self._phase_name[0])
         self._log.debug('<---- End')
 
-    def addExpsDefinition(self, exp_path: str):
+    def addExpDefinitionFromString(self, exp_rcif_content: str) -> NoReturn:
+        self._log.debug('----> Start')
+        """
+        Set cryspy.experiments from a string. Appends to current experiments if available.
+        """
```
Comment (mads-bertelsen) [inline easyInterface/Diffraction/Calculators/CryspyCalculator.py]: I don't believe the documentation will be picked up as a doc string unless it is the first statement in the method, so suggest to start the log after the doc string.

File: easyInterface/Diffraction/Calculators/CryspyCalculator.py [hunk 2]
```diff
@@ -167,14 +185,19 @@ def addExpsDefinition(self, exp_path: str):
             self._log.error('Experiment cif data is malformed')
         if self._cryspy_obj.experiments is not None:
             self._log.info('Adding experiment to cryspy experiments')
+            # Check for the same name and modify if exist
+            if experiment.data_name in self._experiment_name:
+                new_name = experiment.data_name + str(random.randint(0, 100))
+                self._log.warning(f'Experiment {experiment.data_name} exists, renaming to {new_name}')
+                experiment.data_name = new_name
```
Comment (mads-bertelsen) [inline easyInterface/Diffraction/Calculators/CryspyCalculator.py]: I think the random element makes this code prone to randomly fail testing. I suggest adding a while loop inside the current if statement.
```
index = 0
new_name = experiment.data_name
while new_name == self._experiment_name:
    new_name = experiment.data_name + str(index)
    index += 1
```

File: easyInterface/Diffraction/Calculators/CryspyCalculator.py [hunk 3]
```diff
@@ -167,14 +185,19 @@ def addExpsDefinition(self, exp_path: str):
             self._log.error('Experiment cif data is malformed')
         if self._cryspy_obj.experiments is not None:
             self._log.info('Adding experiment to cryspy experiments')
+            # Check for the same name and modify if exist
+            if experiment.data_name in self._experiment_name:
```
Comment (mads-bertelsen) [inline easyInterface/Diffraction/Calculators/CryspyCalculator.py]: Why not use equality instead of "in" here? The new name can be a subset of the old name without causing an issue as far as I understand.

File: easyInterface/Diffraction/Calculators/CryspyCalculator.py [hunk 4]
```diff
@@ -256,14 +279,19 @@ def addPhaseDefinition(self, phases_path: str):
         phase = Crystal().from_cif(phases_rcif_content)
         if self._cryspy_obj.crystals is not None:
             self._log.info('Adding phase to existing phases')
+            # Check for the same name and modify if exist
+            if phase.data_name in self._phase_name:
+                new_name = phase.data_name + str(random.randint(0, 100))
+                self._log.warning(f'Phase {phase.data_name} exists, renaming to {new_name}')
+                phase.data_name = new_name
```
Comment (mads-bertelsen) [inline easyInterface/Diffraction/Calculators/CryspyCalculator.py]: Similar to the other block.
Using equality instead of "in" would lead to less cases with added digits to filenames.
A while loop would be better than random digits.

File: easyInterface/Diffraction/Interface.py [hunk 5]
```diff
@@ -23,7 +23,7 @@ class ProjectDict(LoggedUndoableDict):
     """
 
     def __init__(self, interface: Interface, app: App, calculator: Calculator, info: Info, phases: Phases,
-                 experiments: Experiments, calculations: Calculations) -> None:
+                 experiments: Experiments, calculations: Calculations):
```
Comment (mads-bertelsen) [inline easyInterface/Diffraction/Interface.py]: According to PEP 484 the correct type hint for __init__ is -> None, so why was this removed?

File: easyInterface/Diffraction/Interface.py [hunk 6]
```diff
@@ -42,7 +42,7 @@ def __init__(self, interface: Interface, app: App, calculator: Calculator, info:
         self._log.debug('Created a project dictionary')
 
     @classmethod
-    def default(cls) -> 'LoggedUndoableDict':
+    def default(cls):
```
Comment (mads-bertelsen) [inline easyInterface/Diffraction/Interface.py]: This type hint is less obvious at it is in principle the same as __init__, but uses an explicit return. Why was this hint removed?

File: easyInterface/Diffraction/Interface.py [hunk 7]
```diff
@@ -60,7 +60,7 @@ def default(cls) -> 'LoggedUndoableDict':
     @classmethod
     def fromPars(cls, experiments: Union[Experiments, Experiment, List[Experiment]],
                  phases: Union[Phases, Phase, List[Phase]],
-                 calculations: Optional[Union[Calculations, Calculation, List[Calculation]]] = {}) -> 'LoggedUndoableDict':
+                 calculations: Optional[Union[Calculations, Calculation, List[Calculation]]] = {}):
```
Comment (mads-bertelsen) [inline easyInterface/Diffraction/Interface.py]: Same as above, why was the hint removed?

File: tests/easyInterface/Diffraction/test_Interface.py [hunk 8]
```diff
@@ -93,6 +94,44 @@ def test_setInfoDict(cal):
...
+    assert cal.project_dict['phases']['Fe3O4']['cell']['length_a'].value == 8.36212
+    cal.setPhaseValue('Fe3O4', ['cell', 'length_a'], 5)
+    assert cal.project_dict['phases']['Fe3O4']['cell']['length_a'].value == 5
+
+
+def test_setPhase(cal):
+    phase = cal.getPhase('Fe3O4')
+    cal.project_dict['phases']['Fe3O4']['cell']['length_a'].value = 2
+    cal.setPhase(phase)
+    assert cal.project_dict['phases']['Fe3O4']['cell']['length_a'].value == 8.36212
```
Comment (mads-bertelsen) [inline tests/easyInterface/Diffraction/test_Interface.py]: This function has already been defined, and with the same content, on line 97-101.

File: tests/easyInterface/Diffraction/test_Interface.py [hunk 9]
```diff
@@ -139,30 +178,111 @@ def test_setPhasesDictFromCryspyObj(cal):
...
-    assert phase_dict['Fe3O4']['atoms']['Fe3A']['U_iso_or_equiv'].max == 1.0
+    assert phase_dict['Fe3O4']['atoms']['Fe3A']['U_iso_or_equiv']['store'].max == 1
     assert phase_dict['Fe3O4']['atoms']['O']['U_iso_or_equiv'].value == 0.0
     assert phase_dict['Fe3O4']['atoms']['O']['U_iso_or_equiv']['store']['constraint'] is None
 
     # Isotropic ADP
     assert phase_dict['Fe3O4']['atoms']['Fe3A']['ADP']['u_11']['header'] == 'U11'
     assert phase_dict['Fe3O4']['atoms']['Fe3A']['ADP']['u_11'].value is None
     assert phase_dict['Fe3O4']['atoms']['Fe3A']['ADP']['u_22'].value is None
-    assert phase_dict['Fe3O4']['atoms']['Fe3A']['ADP']['u_23'].min == -np.Inf
```
Comment (mads-bertelsen) [inline tests/easyInterface/Diffraction/test_Interface.py:151]: No replacement for checking the minimum value?

**Review by mads-bertelsen — COMMENTED** [review]
Great to see the functionality necessary for DAS-229 added so fast, only have a few comments for improvements that should be addressed.

File: easyInterface/Diffraction/Calculators/CryspyCalculator.py (hunk 1 above)
Comment (wardsimon) [reply easyInterface/Diffraction/Calculators/CryspyCalculator.py]: Fixed

File: easyInterface/Diffraction/Calculators/CryspyCalculator.py (hunk 2 above)
Comment (wardsimon) [reply easyInterface/Diffraction/Calculators/CryspyCalculator.py]: Yep, non-lazy way has been implemented

File: easyInterface/Diffraction/Calculators/CryspyCalculator.py (hunk 3 above)
Comment (wardsimon) [reply easyInterface/Diffraction/Calculators/CryspyCalculator.py]: `experiment.data_name in self._experiment_name`
str in list. Some bad naming

File: easyInterface/Diffraction/Calculators/CryspyCalculator.py (hunk 4 above)
Comment (wardsimon) [reply easyInterface/Diffraction/Calculators/CryspyCalculator.py]: Fixed

File: easyInterface/Diffraction/Interface.py (hunk 5 above)
Comment (wardsimon) [reply easyInterface/Diffraction/Interface.py]: Fixed

File: easyInterface/Diffraction/Interface.py (hunk 6 above)
Comment (wardsimon) [reply easyInterface/Diffraction/Interface.py]: It wasn't returning a `'LoggedUndoableDict'` replaced with `'ProjectDict'`

File: easyInterface/Diffraction/Interface.py (hunk 7 above)
Comment (wardsimon) [reply easyInterface/Diffraction/Interface.py]: See above

File: tests/easyInterface/Diffraction/test_Interface.py (hunk 8 above)
Comment (wardsimon) [reply tests/easyInterface/Diffraction/test_Interface.py]: Removed. Total face palm

File: tests/easyInterface/Diffraction/test_Interface.py (hunk 9 above)
Comment (wardsimon) [reply tests/easyInterface/Diffraction/test_Interface.py:151]: Removed due to min/max features getting updated. Annoyingly -np.Inf != -np.Inf :-/

File: tests/easyInterface/Diffraction/test_Interface.py [hunk 10]
```diff
@@ -443,63 +563,3 @@ def test_addExperimentDefinition(cal):
     assert exp_added['phase']['Fe3O4']['name'] == exp_ref['phase']['Fe3O4']['name']
     assert exp_added['phase']['Fe3O4']['scale'].value == exp_ref['phase']['Fe3O4']['scale'].value
 
-
```
Comment (wardsimon) [inline tests/easyInterface/Diffraction/test_Interface.py]: The missing tests!

**Review by mads-bertelsen — APPROVED** [review]
Great work 👍
